### PR TITLE
Drop unpublished errata functionality (Web UI)

### PR DESCRIPTION
**Pull request by moio**
### Background

This PR adds patches on top of #2296 to also drop functionality from the UI relative to displaying, looking up and creating unpublished patches. Both PRs will be merged at the same time when approved.

Please refer to #2296 for background information.

### Contents of this PR

 - drop the unpublished patch ACL
 - change the UI for creating a new patch. Before this PR any new patch was first created unpublished and then published in a second step, this patch allows to create published patch directly
 - change the UI for cloning an patch. Before this PR any newly cloned patch was first created unpublished and then published in a second step, this patch allows to clone and publish in one step
 - remove the unpublished patch list link from the menu
 - adapt tests

### Menu
Before: 
image

After:
image

### Creation of patch

Before:
image

After:
image
(top of form is identical until "Notes" - only the channel list is added)

### Cloning of patch (confirm screen)

The selection of patches is identical. After clicking on "Clone" the confirm screen changes per below.

Before:
image

After:
image

- [x] **DONE**

## Documentation

- [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pull/314)

- [x] **DONE**

## Test coverage
- Unit tests were adapted
- Cucumber tests were adapted

- [x] **DONE**

## Links

Partially addressing https://github.com/SUSE/spacewalk/issues/11548

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by moio** [comment]
@ncounter can you please take a look at the UI bits?

**Comment by moio** [comment]
@mcalmer should anything in particular be done about translations?

**Comment by moio** [comment]
> Right about stale things, if the `unpublished` is completely gone, I believe there is much more to remove: in the `struts-config.xml` there are few actions, and some JSPs afterwards like `unpublished.jsp`, `unpublisheddeleteconfirm.jsp`, then few `hbm.xml` files and some `.java` classes too. Do I get it correctly?

Yes you do, actually most of the code hasn't been dropped yet. There will also be at least 6 database tables, all related Hibernate mappings, queries, fragments of stored procedures and so on.

This is only the beginning!

Thanks for the review

**Comment by moio** [comment]
I had a look at unit tests and most are harmless (=expected failures). Only one needs a deeper look and maybe really uncovered a bug. Will fix them all before merging of course.

I will also implement Dario's suggestion on the extra text before proceeding.

As per the Cucumber testsuite, I have one running and will proceed merging before end of day only if results do not suggest further problems.

Thanks

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java [hunk 1]
```diff
@@ -246,18 +246,20 @@ protected void processMethodKeys(Map<String, String> map) {
      * {@inheritDoc}
      * Add eid to our parameter map
      */
-    protected void processParamMap(ActionForm formIn,
+    public void processParamMap(ActionForm formIn,
                                    HttpServletRequest request,
                                    Map<String, Object> params) {
         //keep eid in params
-        Long eid = new RequestContext(request).getRequiredParam("eid");
+        Long eid = new RequestContext(request).getParamAsLong("eid");
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java:253]: makes sense to use com.redhat.rhn.frontend.struts.RequestContext.ERRATA_ID here as well

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java [hunk 2]
```diff
@@ -246,18 +246,20 @@ protected void processMethodKeys(Map<String, String> map) {
...
+        if (eid != null) {
+            params.put("eid", eid);
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java:255]: same here

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelSetupAction.java [hunk 3]
```diff
@@ -65,7 +68,8 @@ public ActionForward execute(ActionMapping mapping,
         clampListBounds(pc, request, user);
 
         //Get the errata object
-        Errata e = requestContext.lookupErratum();
+        Long errataId = requestContext.getParamAsLong(ERRATA_ID);
+        Errata e = ErrataManager.lookupErrata(errataId, requestContext.getCurrentUser());
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/action/errata/ChannelSetupAction.java:72]: might make sense to return an optional here if Errata can be null. Maybe for a next PR

File: java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java [hunk 4]
```diff
@@ -134,6 +148,7 @@ public ActionForward execute(ActionMapping mapping,
         e.setOrg(user.getOrg());
 
         ErrataManager.storeErrata(e);
+        ErrataManager.publish(e, channelAction.getChannelIdsFromRhnSet(set), user);
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java:151]: is the whole concept of "publishing an errata" still valid? can it be an errata in another state that is "not published"?
shouldn't we do this as part of the creation of the errata?

File: java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml [hunk 5]
```diff
@@ -3384,7 +3384,7 @@ button below, and &lt;b&gt;will be unable to log back in&lt;/b&gt;.</source>
         </context-group>
       </trans-unit>
       <trans-unit id="errata.publish.nochannelsselected">
-        <source>A published patch must belong to at least one channel.</source>
+        <source>At least one channel must be selected for publishing.</source>
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:3387]: Should we still use the "publish" word in the UI? to me it gives some sense of "an errata can be in another state than 'published'"

File: java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml [hunk 6]
```diff
@@ -7191,7 +7191,7 @@ organization management tools if you choose to remove this role from your login.
         </context-group>
       </trans-unit>
       <trans-unit id="cloneconfirm.jsp.pagesummary">
-        <source>The patches listed below will be cloned. Once they are cloned, they will appear in your &lt;a href=&quot;{0}&quot;&gt;unpublished patches list&lt;/a&gt;. From there you can verify the text and associated packages for each patch. Once you are ready, you can publish the patches so they are available to users in your organization.</source>
+        <source>The patches listed here will be cloned and published in the channels selected below. Once they are cloned, they will appear in each channel's patch list.</source>
```
Comment (chiaradiamarcelo) [inline java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:7194]: same here. I would remove the the "publish":

"The patches listed here will be cloned in the channels selected below. Once they are cloned, they will appear in each channel's patch list."

**Comment by moio** [comment]
@Bischoff ping as this touches one feature (lightly).

Side note: I did run the testsuite locally with and without patches in this PR and #2296 combined (they will be merged together). There was only one additional failure that looks completely unrelated to the changes here (addition of channels in a CLM project, `Add new sources and promote again`).

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelSetupAction.java (hunk 3 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/action/errata/ChannelSetupAction.java:72]: Yes, I am adding that to the list for the follow-up cleanup PR.

File: java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java (hunk 4 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java:151]: The concept of publishing an errata per se remains - it's just that before, one could publish both an unpublished or a published errata - now unpublished errata cease to exist. So the only remaining option is to re-publish something that was already published before - which is a valid thing to do.

On a high level, I completely agree the "first publishing" should be part of errata creation. Whether that means to wrap those methods together in this specific case is something I would leave for the follow up PR.

So see you there, if you agree.

File: java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml (hunk 5 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:3387]: Reworded.

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java (hunk 1 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java:253]: Addressed

File: java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java (hunk 2 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/action/errata/ChannelAction.java:255]: Addressed

File: java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml (hunk 6 above)
Comment (moio) [reply java/code/src/com/redhat/rhn/frontend/strings/jsp/StringResource_en_US.xml:7194]: Reworded

File: java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java (hunk 4 above)
Comment (chiaradiamarcelo) [reply java/code/src/com/redhat/rhn/frontend/action/errata/CreateAction.java:151]: It would be nice to address this in a follow up PR for clarify. Thanks

**Comment by moio** [comment]
- Unit tests fixed
- Checkstyle issues fixed
- testsuite was run locally with and without patches in this PR and #2296 combined. There was only one additional failure that looks completely unrelated to the patches (a failed addition of channels in a CLM project, `Add new sources and promote again`)
- all other comments were addressed
- PR for documentation added

I am merging this now (and staying available in case of unexpected regressions).

**Pull request merged**